src/main/java/io/cdap/plugin/servicenow/source/ServiceNowMultiRecordReader.java

```diff
@@ -0,0 +1,119 @@
+import io.cdap.plugin.servicenow.source.apiclient.ServiceNowTableAPIClientImpl;
+import io.cdap.plugin.servicenow.source.apiclient.ServiceNowTableDataResponse;
+import io.cdap.plugin.servicenow.source.util.SchemaBuilder;
+import io.cdap.plugin.servicenow.source.util.ServiceNowConstants;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Record reader that reads the entire contents of a ServiceNow table.
+ */
+public class ServiceNowMultiRecordReader extends ServiceNowRecordReader {
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowMultiRecordReader.class);
+  private final ServiceNowMultiSourceConfig multiSourcePluginConf;
+
+  ServiceNowMultiRecordReader(ServiceNowMultiSourceConfig multiSourcePluginConf) {
+    super();
+    this.multiSourcePluginConf = multiSourcePluginConf;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException {
+    try {
+      if (results == null) {
+        fetchData();
+      }
+
+      if (!iterator.hasNext()) {
+        return false;
+      }
+
+      row = iterator.next();
+
+      pos++;
+    } catch (Exception e) {
+      throw new IOException("Exception in nextKeyValue", e);
+    }
+    return true;
+  }
+
+  @Override
+  public StructuredRecord getCurrentValue() throws IOException {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
+    recordBuilder.set(tableNameField, tableName);
+
+    try {
+      for (Schema.Field field : tableFields) {
+        String fieldName = field.getName();
+        Object fieldValue = convertToValue(fieldName, field.getSchema(), row);
+        recordBuilder.set(fieldName, fieldValue);
+      }
+    } catch (Exception e) {
+      throw new IOException("Error decoding row from table " + tableName, e);
+    }
+    return recordBuilder.build();
+  }
+
+  public void fetchData() {
+    tableName = split.getTableName();
+    tableNameField = multiSourcePluginConf.getTableNameField();
+
+    ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(multiSourcePluginConf);
+
+    // Get the table data
+    results =
+      restApi.fetchTableRecords(tableName, multiSourcePluginConf.getStartDate(), multiSourcePluginConf.getEndDate(),
+        split.getOffset(), ServiceNowConstants.PAGE_SIZE);
+
+    LOG.debug("size={}", results.size());
+    if (!results.isEmpty()) {
+      fetchSchema(restApi);
+    }
+
+    iterator = results.iterator();
+  }
+
+  public void fetchSchema(ServiceNowTableAPIClientImpl restApi) {
+    // Fetch the column definition
+    ServiceNowTableDataResponse response = restApi.fetchTableSchema(tableName, null, null,
+      false);
+    if (response == null) {
+      return;
+    }
+
+    // Build schema
+    SchemaBuilder schemaBuilder = new SchemaBuilder();
+    Schema tempSchema = schemaBuilder.constructSchema(tableName, response.getColumns());
```

Comment:
can this just be a static method?

Reply:
I'm sorry I couldn't understand. Which method do we want to be static here?

`private void fetchSchema(ServiceNowTableAPIClientImpl restApi) `

**OR**

`public Schema constructSchema(String tableName, List<ServiceNowColumn> columns) `

If it's the latter, then yes. We can make it static.

Reply:
The latter. I think there's any reason to create an instance of SchemaBuilder - it's just a util class

Reply:
I've made this method: `Schema constructSchema(String tableName, List<ServiceNowColumn> columns)` static and refactored the method call. Now, it's using the class name to directly call the method instead of creating a new instance of that class. [c5d6fbb](https://github.com/data-integrations/servicenow-plugins/pull/8/commits/c5d6fbb1b2d49b43e78aadd471064b92574c868e)